pkg/storegateway/bucket.go

```diff
@@ -1991,38 +1992,33 @@ func toPostingGroup(lvalsFn func(name string) ([]string, error), m *labels.Match
 		}
 	}
 
-	vals, err := lvalsFn(m.Name)
-	if err != nil {
-		return nil, err
-	}
-
 	// This is a more generic approach for the previous case.
 	// Here we can enter with `label=""` or regexp matchers that match the empty value,
 	// like `=~"|foo" or `error!~"5..".
 	// Remember: if the matcher selects an empty value, it selects all the series which don't
 	// have the label name set too. See: https://github.com/prometheus/prometheus/issues/3575
 	// and https://github.com/prometheus/prometheus/pull/3578#issuecomment-351653555.
 	if m.Matches("") {
-		var toRemove []labels.Label
-		for _, val := range vals {
-			if !m.Matches(val) {
-				toRemove = append(toRemove, labels.Label{Name: m.Name, Value: val})
-			}
+		vals, err := lvalsFn(m.Name, not(m.Matches))
+		toRemove := make([]labels.Label, len(vals))
```

Comment:
nit: we could skip this allocation if `err` was non-nil, same below

Reply:
Discussed offline and will leave as is. The error case is very rare here (even though that's not the contract, with current implementation it would only happen if the block is corrupted), so we'd just clutter the code with an extra err check (plus the cpu cost!) for a very rare case.

Reply:
> Makes sense to me. I also agree with Peter's comments.

My comments are non-blocking nits, feel free to ignore them.